src/dmd/cppmanglewin.d

```diff
@@ -669,149 +669,241 @@ private:
+        {
+            sinteger_t val = e.toInteger();
+            if (val < 0)
             {
-                buf.writestring("?A");
-                return;
+                val = -val;
+                tmp.buf.writeByte('?');
             }
-            else if (sym.ident == Id.call)
+            tmp.mangleNumber(val);
+        }
+    }
+
+    void mangleTemplateAlias(RootObject o,VisualCPPMangler tmp)
+    {
+        Dsymbol d = isDsymbol(o);
+        Expression e = isExpression(o);
+        if (!d && !e)
+        {
+            sym.error("Internal Compiler Error: `%s` is unsupported parameter for C++ template", o.toChars());
+            fatal();
+        }
+        if (d && d.isFuncDeclaration())
+        {
+            tmp.buf.writeByte('$');
+            tmp.buf.writeByte('1');
+            tmp.mangleFunction(d.isFuncDeclaration());
+        }
+        else if (e && e.op == TOK.variable && (cast(VarExp)e).var.isVarDeclaration())
+        {
+            tmp.buf.writeByte('$');
+            if (flags & IS_DMC)
+                tmp.buf.writeByte('1');
+            else
+                tmp.buf.writeByte('E');
+            tmp.mangleVariable((cast(VarExp)e).var.isVarDeclaration());
+        }
+        else if (d && d.isTemplateDeclaration() && d.isTemplateDeclaration().onemember)
+        {
+            Dsymbol ds = d.isTemplateDeclaration().onemember;
+            if (flags & IS_DMC)
             {
-                buf.writestring("?R");
-                return;
+                tmp.buf.writeByte('V');
             }
-            else if (sym.ident == Id.cppdtor)
+            else
             {
-                buf.writestring("?_G");
-                return;
+                if (ds.isUnionDeclaration())
+                {
+                    tmp.buf.writeByte('T');
+                }
+                else if (ds.isStructDeclaration())
+                {
+                    tmp.buf.writeByte('U');
+                }
+                else if (ds.isClassDeclaration())
+                {
+                    tmp.buf.writeByte('V');
+                }
+                else
+                {
+                    sym.error("Internal Compiler Error: C++ templates support only integral value, type parameters, alias templates and alias function parameters");
+                    fatal();
+                }
             }
+            tmp.mangleIdent(d);
+        }
+        else
+        {
+            sym.error("Internal Compiler Error: `%s` is unsupported parameter for C++ template: (%s)", o.toChars());
+            fatal();
         }
+    }
+
+    void mangleTemplateType(RootObject o, VisualCPPMangler tmp)
+    {
+        Type t = isType(o);
+        assert(t);
+        t.accept(tmp);
+    }
+
+    void mangleName(Dsymbol sym, bool dont_use_back_reference = false)
+    {
+        //printf("mangleName('%s')\n", sym.toChars());
+        const(char)* name = null;
+        bool is_dmc_template = false;
+
+        if (mangleSpecialName(sym)
```

Comment:
seems to be missing a closing paren

Reply:
I realised that, but I stuff up with git locally so will have to force push, so I'm addressing all these comments now.